features/vat-replacement/run.js

```diff
@@ -0,0 +1,213 @@
+        }
+    }
+
+    try {
+        let taxCategories = []
+        for await (const batch of ctpClient.fetchGraphQlBatches(params)) {
+            await Promise.map(batch, async (taxCategory) => {
+                taxCategories.push(taxCategory)
+            }, { concurrency: 4 })
+        }
+        return taxCategories
+    } catch (e) {
+        console.log(e)
+    }
+}
+
+async function replaceTaxRate(ctpClient, taxRateDraft, taxCategoryItem) {
+    try {
+        const actions = [{
+            action : 'replaceTaxRate',
+            taxRateId : taxRateDraft.id,
+            taxRate : taxRateDraft
+        }]
+        await ctpClient.update(ctpClient.builder.taxCategories, taxCategoryItem.id, taxCategoryItem.version, actions)
+    } catch (e) {
+        console.log(e)
+    }
+}
+
+async function getGermanValidTaxRateList(taxCategories, rateIdToCategoryMap, isDryRun) {
+    let germanTaxRateList = taxCategories
+        .flatMap(item => item.rates)
+        .filter(rate => rate.country=='DE')
+    let validGermanTaxRateList = []
+    if (germanTaxRateList.length==0) {
+        let errMsg  =   'No valid tax rate from Germany. There is nothing to be done in your project in ' +
+                        'respect to VAT.' + '\n'
+        console.error(errMsg)
+        return validGermanTaxRateList
+    }
+    let invalidGermanTaxRateList = germanTaxRateList.filter(rate =>
+        rate.amount != taxRate.standard.original &&
+        rate.amount != taxRate.reduced.original)
+    validGermanTaxRateList = germanTaxRateList
+        .filter(rate => rate.amount==taxRate.standard.original || rate.amount==taxRate.reduced.original)
+
+    const standardVATs = validGermanTaxRateList.filter(rate => rate.amount == taxRate.standard.original)
+    const reducedVATs = validGermanTaxRateList.filter(rate => rate.amount == taxRate.reduced.original)
+    let errMsg  = null
+    if (invalidGermanTaxRateList.length>0) {
+        errMsg  = 'We are sorry, we would have to ask you to change the vat manually if applicable. ' +
+            'There seems to be a special case.' + '\n'
+        errMsg += await buildTaxRateDraftJson(invalidGermanTaxRateList, rateIdToCategoryMap)
+        console.error(errMsg)
+    } else if (standardVATs.length > 1) {
+        errMsg  = 'We are sorry but there are several tax rates for "DE" with a percentage ' +
+            'of ' + taxRate.standard.original*100 + ' percent. Please update the tax rate manually. ' + '\n'
+        errMsg += await buildTaxRateDraftJson(standardVATs, rateIdToCategoryMap)
+        console.error(errMsg)
+    } else if (reducedVATs.length > 1) {
+        errMsg  = 'We are sorry but there are several tax rates for "DE" with a percentage ' +
+            'of ' + taxRate.reduced.original*100 + ' percent. Please update the tax rate manually. ' + '\n'
+        errMsg += await buildTaxRateDraftJson(reducedVATs, rateIdToCategoryMap)
+        console.error(errMsg)
+    }
+
+    return validGermanTaxRateList
```

Comment:
In case of error like `invalidGermanTaxRateList.length>0`, you still return `validGermanTaxRateList` and these will still be processed. According to our task, we have there:
```
if no taxRate for "DE" exists: 'There is nothing to be done in your project in respect to VAT.' –> **no JSON or API call**
if there is a taxRate for "DE" that is NOT .07 or .19 percent set as rate then display this taxRate and the respective taxCategory together with a message like 'We are sorry, we would have to ask you to change the vat manually if applicable. There seems to be a special case.' -> **no JSON or API call**
if there is more than one taxRate for "DE" that is 19 or 7 percent display these taxRates and the respective taxCategories together with a message like 'We are sorry but there are several tax rates for "DE" with a percentage of \variable_with_respective_value\ percent. Please update the taxRate manually' -> **no JSON or API call**
```
Did you clarify this with someone?

Reply:
Just clarified with Thomas, the script suppose to proceed further in case `validGermanTaxRateList` have something in it.